homeassistant/components/epson/__init__.py

```diff
@@ -83,22 +86,42 @@ async def async_migrate_entry(hass, config_entry):
     """Migrate old entry."""
     _LOGGER.debug("Migrating unique_id in %s", config_entry.version)
 
-    old_unique_id = config_entry.unique_id
-    projector = hass.data[DOMAIN][config_entry.entry_id]
-    new_unique_id = await projector.get_serial_number()
-
-    @callback
-    def update_unique_id(entity_entry):
-        """Update unique ID of entity entry."""
-        return {
-            "new_unique_id": entity_entry.unique_id.replace(
-                old_unique_id, new_unique_id
+    async def get_unique_id():
+        try:
+            projector = await validate_projector(hass, config_entry.data[CONF_HOST])
+        except (CannotConnect, PoweredOff):
+            return None
+        else:
+            return await projector.get_serial_number()
+
+    if config_entry.version == 1:
+        data = {CONF_HOST: config_entry.data[CONF_HOST]}
+        hass.config_entries.async_update_entry(config_entry, data=data)
+        config_entry.version = 2
+    if config_entry.version == 2:
+        old_unique_id = config_entry.unique_id
+        new_unique_id = await get_unique_id()
+        registry = await async_get_registry(hass)
+        old_entity_id = registry.async_get_entity_id(
+            "media_player", DOMAIN, config_entry.entry_id
+        )
+        if old_entity_id is not None:
+            registry.async_remove(old_entity_id)
```

Comment:
We should probably not remove the old entry unless we migrate the unique id.

Reply:
as I test it is better to remove old entry as if I leave old one then it creates two entitites `media_player.epson` and `media_player.epson_2` as first one has saved unique_id (which is wrong id - config_entry.entry_id), so I remove it and it media_player then creating entity without unique_id and adding it when unique_id (proper - serial number) become available.

Reply:
Right. I'm saying we should group the remove just before the migrating and not have them in separate places in the code since they are dependent.

Reply:
To clarify again, either we need to update the old entity entry and change its unique id to the new unique_id or we need to update a new entity entry with all the settings (name, icon etc) from the old entity entry.

We can't remove the old entry before we migrate the user customizations for that entry in the entity registry.

If possible I'd recommend to just update the unique_id of the existing entity entry as that's less work.

Reply:
I moved all logic to one place in media_player entity, so it waits till projector is ON.